src/chainstate/stacks/index/trie.rs

```diff
@@ -0,0 +1,2053 @@
+        cursor.retarget(&new_node.clone(), &ret, &storage.get_cur_block());
+
+        trace!("splice_leaf: node-X' at {:?}", &ret);
+        Ok(ret)
+    }
+
+    /// Add a new value to the Trie at the location pointed at by the cursor.
+    /// Returns a ptr to be inserted into the last node visited by the cursor.
+    pub fn add_value(storage: &mut TrieFileStorage, cursor: &mut TrieCursor, value: &mut TrieLeaf) -> Result<TriePtr, Error> {
+        let mut node = match cursor.node() {
+            Some(n) => n,
+            None => panic!("Cursor is uninitialized")
+        };
+
+        if cursor.eop() {
+            match node {
+                TrieNodeType::Leaf(_) => {
+                    return Trie::replace_leaf(storage, cursor, value);
+                },
+                _ => {}
+            };
+
+            Trie::insert_leaf(storage, cursor, value, &mut node)
+        }
+        else {
+            // didn't reach the end of the path, so we're on an intermediate node
+            // or we're somewhere in the path of a leaf.
+            // Either tack the leaf on (possibly promoting the node), or splice the leaf in.
+            if cursor.eonp(&node) {
+                trace!("eop = {}, eonp = {}, c = {:?}, node = {:?}", cursor.eop(), cursor.eonp(&node), cursor, &node);
+                Trie::insert_leaf(storage, cursor, value, &mut node)
+            }
+            else {
+                match node {
+                    TrieNodeType::Leaf(ref mut data) => {
+                        Trie::promote_leaf_to_node4(storage, cursor, data, value)
+                    },
+                    _ => {
+                        Trie::splice_leaf(storage, cursor, value, &mut node)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Calculate the byte vector of the ancestor root hashes of this trie.
+    /// s must point to the block that contains the trie's root.
+    pub fn get_trie_ancestor_hashes_bytes(storage: &mut TrieFileStorage, hash_buf: &mut Vec<u8>) -> Result<(), Error> {
+        let cur_block_header = storage.get_cur_block();
+        let cur_block_rw = storage.readwrite();
+        
+        let mut depth = 0;
+        loop {
+            let prev_block_header = match storage.block_walk(1u32 << depth) {
+                Ok(h) => {
+                    h
+                },
+                Err(e) => {
+                    match e {
+                        Error::NotFoundError => {
+                            // out of blocks
+                            break;
+                        },
+                        _ => {
+                            return Err(e);
+                        }
+                    }
+                }
+            };
+
+            storage.open_block(&prev_block_header, false)?;
+            
+            let root_ptr = TriePtr::new(TrieNodeID::Node256, 0, storage.root_ptr() as u32);
```

Comment:
This construction turns up a lot -- can this be DRY'ed ?

Reply:
Yeah -- I can add a method that generates a TriePtr for the root node.

Reply:
Should be fixed now.